source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,188 @@
+    chunks_.push_back(std::make_unique<CacheInsertChunkBody>(chunk, end_stream));
+    if (!sent_watermark_ && queue_size_bytes_ > high_watermark_bytes_) {
+      over_high_watermark_callback_();
+      sent_watermark_ = true;
+    }
+  } else {
+    insert_context_->insertBody(
+        Buffer::OwnedImpl(chunk),
+        [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+    chunk_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::insertTrailers(const Http::ResponseTrailerMap& trailers) {
+  end_in_sight_ = true;
+  if (chunk_in_flight_) {
+    chunks_.push_back(std::make_unique<CacheInsertChunkTrailers>(trailers));
+  } else {
+    insert_context_->insertTrailers(trailers,
+                                    [this](bool ready) { onChunkComplete(ready, true, 0); });
+    chunk_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::onChunkComplete(bool ready, bool end_stream, size_t sz) {
+  // If the cache implementation is asynchronous, this may be called from whatever
+  // thread that cache implementation runs on. Therefore, we post it to the
+  // dispatcher to be certain any callbacks and updates are called on the filter's
+  // thread (and therefore we don't have to mutex-guard anything).
+  dispatcher_.post([this, ready, end_stream, sz]() {
+    chunk_in_flight_ = false;
+    if (aborting_) {
+      // Parent filter was destroyed, so we can quit this operation.
+      self_ownership_.reset();
+      return;
+    }
+    ASSERT(queue_size_bytes_ >= sz, "queue can't be emptied by more than its size");
+    queue_size_bytes_ -= sz;
+    if (sent_watermark_ && queue_size_bytes_ <= low_watermark_bytes_) {
+      under_low_watermark_callback_();
+      sent_watermark_ = false;
+    }
+    if (!ready) {
+      // canceled by cache; unwatermark if necessary, inform the filter if
+      // it's still around, and delete the queue.
+      if (sent_watermark_) {
+        under_low_watermark_callback_();
+      }
+      chunks_.clear();
+      self_ownership_.reset();
+      abort_callback_();
+      return;
+    }
+    if (end_stream) {
+      ASSERT(chunks_.empty(), "ending a stream with the queue not empty is a bug");
+      ASSERT(!sent_watermark_,
+             "being over the high watermark when the queue is empty makes no sense");
+      self_ownership_.reset();
+      return;
+    }
+    if (!chunks_.empty()) {
+      // If there's more in the queue, push the next chunk to the cache.
+      auto chunk = std::move(chunks_.front());
+      chunks_.pop_front();
+      chunk_in_flight_ = true;
+      chunk->send(*insert_context_, [this](bool ready, bool end_stream, size_t sz) {
+        onChunkComplete(ready, end_stream, sz);
+      });
+    }
+  });
+}
+
+void CacheInsertQueue::takeOwnershipOfYourself(std::unique_ptr<CacheInsertQueue> self) {
+  // Disable all the callbacks, they're going to have nowhere to go.
+  abort_callback_ = under_low_watermark_callback_ = over_high_watermark_callback_ = []() {};
```

Comment:
here you could just reset your `OptRef<Http::StreamEncoderFilterCallbacks>` and check that for has_value before dereferencing it.